### DISCUSSION TITLE
Welcome to Indigo Discussions!

### DISCUSSION THREAD

**Discussion by davesmith00000** [PurpleKingdomGames/indigo#85] [General]
## Welcome!
  We’re using Discussions as a place to connect with other members of our community. Please feel free to:
  * Ask questions!
  * Share ideas, feedback, or games and demos you've built!
  * Engage with other community members! (Please be welcoming, respectful, and kind)

**Comment by ororlevy** [comment]
Hey, I'm trying to use your engine, but I've struggled with working on the IDEA framework.  
For some reason, I can't go to the declarations from your engine. Did someone have experience setting up a project on intellij with your project?

**Reply by davesmith00000** [reply]
Hello!

I've made a lot of use of type aliases to make the imports a bit more magical. This used to work fine in Scala 2, but the Scala 3 tooling (Intellij / metals) is struggling with it I'm afraid, and I'm not at the point of taking them out yet because frankly, I'm not sure what that would do to the codebase. Also I don't know if it's just the aliases causing the problem.

For the record I'm using VSCode with Metals and sbt BSP. That's not a glowing review it's the least worst option for me, and no, clicking through still isn't reliable.

No easy answers, but I was discussing this with someone else recently and there were two outcomes:

1. I published the [API docs](https://purplekingdomgames.github.io/indigo/api/indigo.html)! Not ideal but revamped due to Scala 3, and actually I've found the search facility surprisingly useful.
2. If you're using sbt (might be possible with Mill too) and you're not precious about Scala 3, you can use sbt's `.cross(..)` lib dep methods to make Scala 3 libs work with Scala 2 (thanks to TASTy), and apparently that works. I have not tried it myself but I know someone else is working that way.

Hope that helps a bit, sorry I don't have a simple solution for you.

Dave

**Reply by ororlevy** [reply]
Firstly, thanks for your quick answer! 
Secondly, I will try to work with the cross option. The API is great but a bit hard to work with during development.  Maybe I will try to work on a different framework cause, for now, not be able to go through your library (on intellij) and understand each class\method is not convenient and not easy to ramp up.

**Reply by davesmith00000** [reply]
I agree it's far from ideal.

I've met two camps of people so far. There's the people who have been using Scala 3 for ages now, like me, and they're just sort of used to it. And everyone else thinks it's totally unusable! :-)

Let me know how you get on with the sbt cross stuff:

```scala
// build.sbt
lazy val hello = project.in(file("."))
  .settings(
    scalaVersion := "2.13.5",
    scalacOptions += "-Ytasty-reader",
    libraryDependencies +=
      ("org.typelevel" %% "cats-core" % "x.y.z").cross(CrossVersion.for2_13Use3)
  )
```
https://www.scala-lang.org/blog/2021/04/08/scala-3-in-sbt.html

I'll try and have a go myself later to see if I can issue any official advice / workarounds, as it's a common complaint and blocker to adoption.

Whether you persist or not, thanks for bringing it up. :-)

**Reply by ororlevy** [reply]
I've tried to play a bit with the cross-version and not succussed so far :| 
The only thing that works for me right now is to use indigo version 0.7.1 with scala 2.13.5. So I will guess I will use the older version to get things going and understand the basics. 
Did you mean to use the cross option for indigo imports or other imports?
`("io.indigoengine" %%% "indigo-json-circe" % "0.8.2").cross(CrossVersion.for2_13Use3)`
I'm a but cluessless with this options and maybe misunderstood you

**Reply by davesmith00000** [reply]
@ororlevy Today... is your lucky day. It seems.
Metals 0.10.5 has just been released. I've updated my local indigo to use Scala 3.0.1, switched from sbt BSP to Bloop and... go to declaration/definition works!
I've been using it for fully 3 minutes so who knows what problems may yet emerge but that's a good start!

**Reply by ororlevy** [reply]
Hey @davesmith00000. 
It's still not working on the IntelliJ side, even with BSP instead of sbt.  
But it's working on VS code with metals, so I think I will try to work with it or the older version with IntelliJ. 
Thank you very much for your help. I will update if I will find a way to get this working with IntelliJ.

**Reply by kpoupon** [reply]
Hello !
I just got a notification from IntelliJ telling me Scala 3 support is still in development and that I may be interested in using nightly builds. So I went to Settings > Languages & Frameworks > Scala > Updates > Update channel : Nightly Builds.
And for now "go to" actions and syntax highlighting for Indigo objects works fine !
